signer/xpi/x509.go

```diff
@@ -56,30 +56,45 @@ func (s *XPISigner) monitorRsaCacheSize() {
 	}
 }
 
-// retrieve a key from the cache or generate one if it takes too long
-// or if the size is wrong
+var currentRsaKeyAge time.Time
+var currentRsaKeyUsage int
+var currentRsaKey *rsa.PrivateKey
+
+// getRsaKey applies some intelligence to key management. It will return the
+// current key if its lifetime or usage count aren't exceeded, or will get a
+// new key from the cache. If the cache is unresponsive, it will generate a new
+// RSA key by itself and return it. All while publishing stats so we can
+// monitor the state of the cache.
 func (s *XPISigner) getRsaKey(size int) (*rsa.PrivateKey, error) {
 	var (
-		err error
-		key *rsa.PrivateKey
+		err   error
 		start time.Time
 	)
+	if currentRsaKeyAge.Add(s.rsaKeyMaxAge).After(time.Now()) && // if current key hasn't reached max lifetime
+		currentRsaKeyUsage < s.rsaKeyMaxUsage { // if current key hasn't reached max usage
+		currentRsaKeyUsage++
```

Comment:
for multiple concurrent requests will this be atomic?

Reply:
we can test this with https://golang.org/doc/articles/race_detector.html#Race_on_loop_counter